File: apps/web/src/lib/rate-limit.ts
import { NextRequest } from "next/server";
import { ipAddress } from "@vercel/functions";
import { Ratelimit } from "@upstash/ratelimit";
import kv from "@/lib/kv";
import logger from "@/lib/logger";

// Rate limiting configuration: 80 requests per minute sliding window
const ratelimit = new Ratelimit({
  redis: kv,
  limiter: Ratelimit.slidingWindow(80, "1 m"),
});

// Expanded regex to detect various bot and script user agents
const suspiciousUARegex = /python-requests|python\/|requests\/|curl\/|wget\/|go-http-client\/|httpie\/|postman\/|axios\/|node-fetch\/|empty|unknown|bot|crawl|spider/i;

// Store blocked IPs with a TTL
async function isIPBlocked(ip: string): Promise<boolean> {
  const blockedUntil = await kv.get(`blocked:${ip}`);
  return blockedUntil !== null && parseInt(blockedUntil as string) > Date.now();
}

async function blockIP(ip: string, durationHours = 24): Promise<void> {
  const blockUntil = Date.now() + (durationHours * 60 * 60 * 1000);
  await kv.set(`blocked:${ip}`, blockUntil.toString(), { ex: durationHours * 60 * 60 });
  logger.warn(`IP blocked for ${durationHours} hours`, { ip });
}

export interface RateLimitResult {
  success: boolean;
  limit: number;
  reset: number;
  remaining: number;
  error?: string;
}

export async function checkRateLimit(request: NextRequest): Promise<RateLimitResult> {
  const ip = ipAddress(request) ?? "127.0.0.1";
  const ua = request.headers.get("user-agent")?.toLowerCase() || "unknown";

  // Check if IP is already blocked (commented out for now as in original middleware)
  // if (await isIPBlocked(ip)) {
  //   logger.warn("Request from blocked IP", { ip, ua });
  //   return {
  //     success: false,
  //     limit: 0,
  //     reset: 0,
  //     remaining: 0,
  //     error: "IP blocked"
  //   };
  // }

  // Perform rate limiting
  const { success, limit, reset, remaining } = await ratelimit.limit(ip);

  if (!success) {
    logger.warn("Rate limit exceeded", {
      ip,
      limit,
      reset,
      remaining,
      ua,
    });
    return {
      success: false,
      limit,
      reset,
      remaining,
      error: "Rate limit exceeded"
    };
  }

  // Log request details
  logger.info("Request received", {
    ip,
    ua,
    timestamp: Date.now(),
    path: request.nextUrl.pathname,
  });

  // Check for suspicious user agents (warning only, not blocking as in original)
  if (suspiciousUARegex.test(ua)) {
    logger.warn("Suspicious user agent detected", { ip, ua });
    // Could implement IP blocking here if needed:
    // await blockIP(ip, 24);
  }

  // Log warning if approaching rate limit
  if (remaining < 20) {
    logger.warn("Approaching rate limit", {
      ip,
      remaining,
      ua,
    });
  }

  return {
    success: true,
    limit,
    reset,
    remaining
  };
}

export { blockIP, isIPBlocked };

File: apps/api/internal/app/rate_limit.go
package app

import (
	"context"
	"fmt"
	"net"
	"net/http"
	"regexp"
	"strings"
	"sync/atomic"
	"time"

	redis "github.com/redis/go-redis/v9"
)

const (
	rateLimitWindow = 60 * time.Second
	rateLimitCount  = int64(80)
)

var suspiciousUA = regexp.MustCompile(`python-requests|python/|requests/|curl/|wget/|go-http-client/|httpie/|postman/|axios/|node-fetch/|empty|unknown|bot|crawl|spider`)

type RateLimitResult struct {
	Success   bool
	Limit     int64
	Reset     int64
	Remaining int64
	Error     string
}

type RateLimiter struct {
	redis   *redis.Client
	log     *Logger
	counter uint64
}

func NewRateLimiter(redisClient *redis.Client, log *Logger) *RateLimiter {
	return &RateLimiter{redis: redisClient, log: log}
}

func (l *RateLimiter) Check(ctx context.Context, r *http.Request) RateLimitResult {
	ip := clientIP(r)
	ua := strings.ToLower(strings.TrimSpace(r.Header.Get("User-Agent")))
	if ua == "" {
		ua = "unknown"
	}

	now := time.Now()
	key := "ratelimit:" + ip
	cutoff := now.Add(-rateLimitWindow).UnixMilli()
	reset := now.Add(rateLimitWindow).UnixMilli()

	var countCmd *redis.IntCmd
	_, err := l.redis.Pipelined(ctx, func(pipe redis.Pipeliner) error {
		pipe.ZRemRangeByScore(ctx, key, "-inf", fmt.Sprintf("%d", cutoff))
		countCmd = pipe.ZCard(ctx, key)
		return nil
	})
	if err != nil {
		l.log.Warn("Rate limit cleanup failed", map[string]any{"ip": ip, "error": err.Error()})
		return RateLimitResult{Success: true, Limit: rateLimitCount, Remaining: rateLimitCount, Reset: reset}
	}

	count, err := countCmd.Result()
	if err != nil {
		l.log.Warn("Rate limit count failed", map[string]any{"ip": ip, "error": err.Error()})
		return RateLimitResult{Success: true, Limit: rateLimitCount, Remaining: rateLimitCount, Reset: reset}
	}

	if count >= rateLimitCount {
		l.log.Warn("Rate limit exceeded", map[string]any{"ip": ip, "ua": ua, "limit": rateLimitCount, "remaining": 0})
		return RateLimitResult{
			Success:   false,
			Limit:     rateLimitCount,
			Remaining: 0,
			Reset:     reset,
			Error:     "Rate limit exceeded",
		}
	}

	member := fmt.Sprintf("%d-%d", now.UnixMilli(), atomic.AddUint64(&l.counter, 1))
	_, err = l.redis.Pipelined(ctx, func(pipe redis.Pipeliner) error {
		pipe.ZAdd(ctx, key, redis.Z{Score: float64(now.UnixMilli()), Member: member})
		pipe.Expire(ctx, key, rateLimitWindow)
		return nil
	})
	if err != nil {
		l.log.Warn("Rate limit add failed", map[string]any{"ip": ip, "error": err.Error()})
		return RateLimitResult{Success: true, Limit: rateLimitCount, Remaining: rateLimitCount, Reset: reset}
	}

	remaining := rateLimitCount - (count + 1)
	l.log.Info("Request received", map[string]any{"ip": ip, "ua": ua, "path": r.URL.Path, "timestamp": now.UnixMilli()})

	if suspiciousUA.MatchString(ua) {
		l.log.Warn("Suspicious user agent detected", map[string]any{"ip": ip, "ua": ua})
	}

	if remaining < 20 {
		l.log.Warn("Approaching rate limit", map[string]any{"ip": ip, "ua": ua, "remaining": remaining})
	}

	return RateLimitResult{
		Success:   true,
		Limit:     rateLimitCount,
		Remaining: remaining,
		Reset:     reset,
	}
}

func clientIP(r *http.Request) string {
	if forwarded := strings.TrimSpace(r.Header.Get("CF-Connecting-IP")); forwarded != "" {
		return forwarded
	}

	if forwarded := strings.TrimSpace(r.Header.Get("X-Real-IP")); forwarded != "" {
		return forwarded
	}

	if forwarded := strings.TrimSpace(r.Header.Get("X-Forwarded-For")); forwarded != "" {
		parts := strings.Split(forwarded, ",")
		if len(parts) > 0 {
			return strings.TrimSpace(parts[0])
		}
	}

	host, _, err := net.SplitHostPort(strings.TrimSpace(r.RemoteAddr))
	if err == nil && host != "" {
		return host
	}

	return "127.0.0.1"
}
